sklearn/mixture/tests/test_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,257 @@
+    Easily separable two gaussian problem.
+    """
+    np.random.seed(1)
+
+    n = 100
+    d = 3
+
+    X1 = np.random.normal(2, 0.5, size=(n, d))
+    X2 = np.random.normal(-2, 0.5, size=(n, d))
+    X = np.vstack((X1, X2))
+    y = np.repeat([0, 1], n)
+
+    # test BIC
+    gmIC = GaussianMixtureIC(max_components=5, criterion="bic", n_jobs=2)
+    gmIC.fit(X, y)
+    n_components = gmIC.n_components_
+
+    # Assert that the two cluster model is the best
+    assert_equal(n_components, 2)
+
+    # Assert that we get perfect clustering
+    ari = adjusted_rand_score(y, gmIC.fit_predict(X))
+    assert_allclose(ari, 1)
+
+    # test AIC
+    gmIC = GaussianMixtureIC(max_components=5, criterion="aic", n_jobs=2)
+    gmIC.fit(X, y)
+    n_components = gmIC.n_components_
+
+    # AIC gets the number of components wrong
+    assert_equal(n_components >= 1, True)
+    assert_equal(n_components <= 5, True)
+
+
+def test_five_class():
+    """
+    Easily separable five gaussian problem.
+    """
+    np.random.seed(1)
+
+    n = 100
+    mus = [[i * 5, 0] for i in range(5)]
+    cov = np.eye(2)  # balls
+
+    X = np.vstack([np.random.multivariate_normal(mu, cov, n) for mu in mus])
+
+    # test BIC
+    gmIC = GaussianMixtureIC(
+        min_components=3,
+        max_components=10,
+        covariance_type="all",
+        criterion="bic",
+    )
+    gmIC.fit(X)
+    assert_equal(gmIC.n_components_, 5)
+
+    # test AIC
+    gmIC = GaussianMixtureIC(
+        min_components=3,
+        max_components=10,
+        covariance_type="all",
+        criterion="aic",
+    )
+    gmIC.fit(X)
+    # AIC fails often so there is no assertion here
+    assert_equal(gmIC.n_components_ >= 3, True)
+    assert_equal(gmIC.n_components_ <= 10, True)
+
+
+# Generate random data for each of the four covariance types
+n = 100
+mu1 = [-10, 0]
+mu2 = [10, 0]
+
+# Spherical
+np.random.seed(1)
+cov1 = 2 * np.eye(2)
+cov2 = 2 * np.eye(2)
+X1 = np.random.multivariate_normal(mu1, cov1, n)
+X2 = np.random.multivariate_normal(mu2, cov2, n)
+X_spherical = np.concatenate((X1, X2))
+
+# Diagonal
+np.random.seed(10)
+cov1 = np.diag([1, 1])
+cov2 = np.diag([2, 1])
+X1 = np.random.multivariate_normal(mu1, cov1, n)
+X2 = np.random.multivariate_normal(mu2, cov2, n)
+X_diag = np.concatenate((X1, X2))
+
+# Tied
+cov1 = np.array([[2, 1], [1, 2]])
+cov2 = np.array([[2, 1], [1, 2]])
+X1 = np.random.multivariate_normal(mu1, cov1, n)
+X2 = np.random.multivariate_normal(mu2, cov2, n)
+X_tied = np.concatenate((X1, X2))
+
+# Full
+cov1 = np.array([[2, -1], [-1, 2]])
+cov2 = np.array([[2, 1], [1, 2]])
+X1 = np.random.multivariate_normal(mu1, cov1, n)
+X2 = np.random.multivariate_normal(mu2, cov2, n)
+X_full = np.concatenate((X1, X2))
+
+
+@pytest.mark.parametrize(
+    "test_data,expected_cov_type",
+    [
+        (X_spherical, "spherical"),
+        (X_diag, "diag"),
+        (X_tied, "tied"),
+        (X_full, "full"),
+    ],
+)
+def test_covariances(test_data, expected_cov_type):
```

Comment:
The best here for maintainability and to avoid code duplication and variable sitting in the general scope would be:
 - to have `n`, and the two means `mu1` and `mu2` be defined inside the test
 - to parametrise tests on the two covariances (hence `cov1, cov2, expected_cov_type` for the parametrisation)
 - to have `X1`, `X2` and `X` computed inside the test